### PR TITLE
Gen lepton with > 6 mothers

### PR DESCRIPTION
**Pull request by msessini**
This PR includes a warning for gen lepton with more than 6 mothers and store only the 6 firsts.

The cms::Exception is kept for safety but shouldn't occur anymore.

### CODE REVIEW

File: Analysis/interface/GenLepton.h [hunk 1]
```diff
@@ -243,8 +243,15 @@ class GenLepton {
             size_t last_mother_index = NoneIndex;
 
             if(!particle->motherRefVector().empty()) {
-                for(const auto& mother : particle->motherRefVector())
+		int Nmother = 0;
```
Comment (kandrosov) [inline Analysis/interface/GenLepton.h:246]: I think it is better to fix the limitation at the level of filling TauTuple and keep `GenLepton.h` logic intact.
https://github.com/cms-tau-pog/TauMLTools/blob/master/Production/plugins/TauTupleProducer.cc#L401-L420

File: Analysis/interface/GenLepton.h [hunk 2]
```diff
@@ -243,8 +243,15 @@ class GenLepton {
...
+                for(const auto& mother : particle->motherRefVector()) {
                     FillDaughters(mother.get(), NoneIndex, false);
+		    ++Nmother;
+		    if(Nmother>6) {
+			    std::cout<<"Warning : Gen particle with > 6 mothers, storing only 6 firsts"<<std::endl;
+			    break;
+		    }
```
Comment (shedprog) [inline Analysis/interface/GenLepton.h:253]: May I ask: in which MC it is observed? Is it for pp interaction level summing up over all the partons? I am a bit concerned why do we want to save it

Comment (kandrosov) [reply Analysis/interface/GenLepton.h:253]: Having more than 2 mothers occur regularly in all our MC in tau decays from FSR. The total number of mothers in fragmentation does not have an upper limit (from physics perspective). For technical reasons, we can save only up to 6 mothers. But we don't need to have the precise decay graph for FSR products, so I think it is ok to store only first 6 mothers.

File: Analysis/interface/GenLepton.h [hunk 3]
```diff
@@ -243,6 +243,7 @@ class GenLepton {
             size_t last_mother_index = NoneIndex;
 
             if(!particle->motherRefVector().empty()) {
+		int Nmother = 0;
```
Comment (kandrosov) [inline Analysis/interface/GenLepton.h:246]: `Nmother` is not used -> remove

File: Production/plugins/TauTupleProducer.cc [hunk 4]
```diff
@@ -420,6 +425,8 @@ class TauTupleProducer : public edm::EDAnalyzer {
             };
 
             tauTuple().genLepton_lastMotherIndex = static_cast<int>(genLepton->mothers().size()) - 1;
+	    if(genLepton->mothers().size()>6) 
+		    tauTuple().genLepton_lastMotherIndex = 5;
```
Comment (kandrosov) [inline Production/plugins/TauTupleProducer.cc:429]:
```suggestion
            tauTuple().genLepton_lastMotherIndex = static_cast<int>(std::min(genLepton->mothers().size(), 6)) - 1;
```

File: Production/plugins/TauTupleProducer.cc [hunk 5]
```diff
@@ -403,15 +403,20 @@ class TauTupleProducer : public edm::EDAnalyzer {
...
                 Long64_t pos = 0;
                 Long64_t shift = 1;
                 std::set<int> mother_indices;
-                for(auto mother : mothers)
+		int Nmother = 0;
+                for(auto mother : mothers) {
+		    if(Nmother>6) break;
                     mother_indices.insert(getIndex(mother));
+		    ++Nmother;
+		}
```
Comment (kandrosov) [inline Production/plugins/TauTupleProducer.cc:419]:
```suggestion
                for(auto mother : mothers) {
		    if(mother_indices.size() >= 6) break;
                    mother_indices.insert(getIndex(mother));
		}
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Review by kandrosov — APPROVED** [review]

**Review by ofivite — APPROVED** [review]

**Review by shedprog — APPROVED** [review]

**Pull request merged**